zds/tutorialv2/models/models_database.py

```diff
@@ -855,6 +860,175 @@ def get_nb_letters(self, md_file_path=None):
+        m.field('categories', Keyword(boost=2.25))
+        m.field('text', Text())  # for article and mini-tuto, text is directly included into the main object
+
+        return m
+
+    @classmethod
+    def get_es_django_indexable(cls, force_reindexing=False):
+        """Overridden to remove must_redirect=True (and prefetch stuffs).
+        """
+
+        q = super(PublishedContent, cls).get_es_django_indexable(force_reindexing)
+        return q.prefetch_related('content')\
+            .prefetch_related('content__tags')\
+            .prefetch_related('content__subcategory')\
+            .filter(must_redirect=False)
+
+    @classmethod
+    def get_es_indexable(cls, force_reindexing=False):
+        """Overridden to include chapters as well
+        """
+
+        published_contents = super(PublishedContent, cls).get_es_indexable(force_reindexing)
+        indexable = []
+        chapters = []
+
+        for content in published_contents:
+            versioned = content.load_public_version()
+
+            if versioned.has_sub_containers():  # chapters are only indexed for middle and big tuto
+                for chapter in versioned.get_list_of_chapters():
+                    chapters.append(FakeChapter(chapter, versioned, content.es_id))
+
+        indexable.extend(chapters)
+        indexable.extend(published_contents)
+        return indexable
+
+    def get_es_document_source(self, excluded_fields=None):
+        """Overridden to handle the fact that most information are versioned
+        """
+
+        excluded_fields = excluded_fields or []
+        excluded_fields.extend(['title', 'description', 'tags', 'categories', 'text'])
+
+        data = super(PublishedContent, self).get_es_document_source(excluded_fields=excluded_fields)
+
+        # fetch versioned information
+        versioned = self.load_public_version()
+
+        data['title'] = versioned.title
+        data['description'] = versioned.description
+        data['tags'] = [tag.title for tag in versioned.tags.all()]
+
+        categories = []
+        for subcategory in versioned.subcategory.all():
+            categories.extend([subcategory.title, subcategory.get_parent_category().title])
+        data['categories'] = list(set(categories))  # remove duplicates
+
+        if versioned.has_extracts():
+            data['text'] = versioned.get_content_online()
+
+        return data
+
+
+@receiver(pre_delete, sender=PublishedContent)
+def delete_published_content_in_elasticsearch(sender, instance, **kwargs):
+    """Catch the pre_delete signal to ensure the deletion in ES. Also, handle the deletion of the corresponding
+    chapters.
+
+    For that, it uses the ``_delete_by_query`` API,
+    (https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-delete-by-query.html)
+    which is not available in elasticsearch python API (thought the documentation says the opposite), so it passes
+    directly trough the connection layer.
```

Comment:
I have so many problem with these ones. Also, this docstring not true anymore :-°